### PR TITLE
Full vision pipeline for multiple objects!

### PR DESCRIPTION
**Pull request by nkumar-bdai**
This PR enables us to run our pipeline with perception directly from vision without making the hard assumption that there is only one object in the world (i.e, the brush). We can now easily modify the numbers and types of objects!

Example Command:
```
python predicators/main.py --env spot_bike_env --approach spot_wrapper[oracle] --debug --strips_learner oracle --sampler_learner oracle --bilevel_plan_without_sim True --num_test_tasks 1 --explorer active_sampler --spot_robot_ip 10.17.30.30 --spot_grasp_use_sam True --perceiver spot_bike_env --execution_monitor expected_atoms --horizon 15 --spot_initialize_surfaces_to_default False --spot_visualize_vision_model_outputs False --seed 123 --test_task_json_dir predicators/envs/assets/task_jsons/spot_bike_env/brush_screwdriver_perception/
```

### CODE REVIEW

File: predicators/spot_utils/perception_utils.py [hunk 1]
```diff
@@ -137,17 +149,20 @@ def query_detic_sam(image_in: np.ndarray, classes: List[str],
...
-        "boxes": [],
-        "classes": [],
-        "masks": [],
-        "scores": []
-    }
-    for key, value in d.items():
-        d_filtered[key].append(value[selected_idx])
+    for obj_class in classes:
+        class_mask = (d['classes'] == obj_class)
+        if np.all(class_mask == False):  # pylint:disable=singleton-comparison
```
Comment (tsilver-bdai) [inline predicators/spot_utils/perception_utils.py:154]: any reason to not do `not np.any(class_mask)` (and avoid the pylint)?

File: predicators/spot_utils/perception_utils.py [hunk 2]
```diff
@@ -137,17 +149,20 @@ def query_detic_sam(image_in: np.ndarray, classes: List[str],
...
+            continue
+        max_score = np.max(d['scores'][class_mask])
+        max_score_idx = np.where(d['scores'] == max_score)[0]
+        if d['scores'][max_score_idx] < CFG.spot_vision_detection_threshold:
+            continue
+        for key, value in d.items():
+            # Sanity check to ensure that we're selecting a value from
+            # the class we're looking for.
+            if key == "classes":
```
Comment (tsilver-bdai) [inline predicators/spot_utils/perception_utils.py:163]: I'm having trouble understanding this, maybe will ask you in person

**Review by tsilver-bdai — COMMENTED** [review]
halfway through, looking good so far!

File: predicators/spot_utils/perception_utils.py [hunk 3]
```diff
@@ -377,18 +387,23 @@ def get_object_locations_with_detic_sam(
...
-    # Detect multiple objects with their masks
-    obj_num = len(res_segment['masks'])
-    res_locations = []
-    for i in range(obj_num):
+    if len(res_segment['classes']) == 0:
+        return {}
+
+    ret_obj_positions: Dict[str, Tuple[float, float, float]] = {}
+    for i, obj_class in enumerate(res_segment['classes']):
+        # Check that this particular class is one of the
```
Comment (tsilver-bdai) [inline predicators/spot_utils/perception_utils.py:396]: is it possible that these assertions will crash? should we do something soft that doesn't crash instead?

File: predicators/spot_utils/perception_utils.py [hunk 4]
```diff
@@ -311,34 +326,29 @@ def get_xyz_from_depth(image_response: bosdyn.api.image_pb2.ImageResponse,
 
 
 def get_pixel_locations_with_detic_sam(
-        classes: List[str],
+        obj_class: str,
         in_res_image: Dict[str, np.ndarray],
```
Comment (tsilver-bdai) [inline predicators/spot_utils/perception_utils.py:330]: `np.ndarray` types should be `NDArray`

**Review by tsilver-bdai — APPROVED** [review]
great work!

File: predicators/spot_utils/perception_utils.py (hunk 3 above)
Comment (nkumar-bdai) [reply predicators/spot_utils/perception_utils.py:396]: no - it shouldn't be possible

**Pull request merged**